api/envoy/api/v2/lds.proto

```diff
@@ -131,6 +131,12 @@ message Listener {
   // timeout. If not specified, a default timeout of 15s is used.
   google.protobuf.Duration listener_filters_timeout = 15 [(gogoproto.stdduration) = true];
 
+  // [#not-implemented-hide:]
+  // QUIC listener for handing UDP packets. Only used when :ref:`address
+  // <envoy_api_field_Listener.address>` specifies a UDP :ref:`SocketAddress
+  // <envoy_api_msg_core_SocketAddress>`.
+  listener.QuicListener quic_listener = 16;
```

Comment:
Unfortunately we are in a bit of an API bind here. If I were starting from scratch, I think I would have different listener types with possible a CommonListener section shared by all the types, so I think this is going to get ugly no matter what we do, as we are going to have a lot of special cases that look like "if you have this listener type, these fields don't apply." 

With that said, I don't think it's worth going through the deprecation dance to fix this as it would be a huge change. If people think we should do that though, I think we should discuss it sooner rather than later (cc @htuch for API thoughts). For the rest of this review I will assume we aren't going to do a fundamental refactoring.

On this particular topic, IMO, I think we can do basic UDP without a factory. If it's connection oriented most of the fields apply, and if it's not connection oriented, we can likely have a new type of filter chain that instantiates static filters (maybe the existing filter chain config can still be used w/ runtime checks for filter types). So I guess we will need some way for basic UDP of pivoting between connection oriented and static at the config level. @conqerAtapple thoughts here?

So on the QUIC topic, if we are committed to supporting runtime QUIC listener loading, then this approach seems fine to me. I agree with @alyssawilk that I would call this `QuicListenerConfig` even if a bit redundant.

Reply:
I guess there are two separate considerations; from an implementation and API perspective. Given that what is proposed in this PR is just to add a new kind of opaque config for pluggable listener variants, it seems we could easily reuse this for UDP etc if we name it correctly. How this manifest in the concrete implementation, `¯\_(ツ)_/¯
`

Reply:
I think `UDP` listener would make more sense here and would work well with #5473